python_modules/dagster/dagster/_core/definitions/assets_job.py

```diff
@@ -267,6 +269,27 @@ def asset2(asset1):
     )
 
 
+def check_for_executable_upstream_of_nonexecutable(assets) -> None:
+    assets_defs_by_key = {}
+    for assets_def in assets:
+        for key in assets_def.keys:
+            assets_defs_by_key[key] = assets_def
+
+    for assets_def in assets:
+        for key in assets_def.keys:
+            if assets_defs_by_key[key].is_executable:
+                continue
+
+            # we are now not executable. Disallowing upstream dependencies on executable
+            for upstream_of_nonexecutable_key in assets_def.asset_deps[key]:
+                if assets_defs_by_key[upstream_of_nonexecutable_key].is_executable:
+                    raise DagsterInvariantViolationError(
+                        "An executable asset cannot be upstream of a non-executable asset."
+                        f' Non-executable asset "{key.to_user_string()}" downstream of executable'
+                        f' asset "{upstream_of_nonexecutable_key.to_user_string()}"'
+                    )
+
```

Comment:
related to https://github.com/dagster-io/dagster/pull/16586#issuecomment-1724511225

hopefully we get to a point where unexecutable assets are just not put in to *any* jobs, or if not that far still something which changes the nature of what we are trying to enforce here